src/main/java/net/neoforged/neoforge/event/entity/EntityTeleportEvent.java

```diff
@@ -29,43 +32,86 @@
  * All children of this event are fired on the {@link NeoForge#EVENT_BUS}.<br>
  **/
 public class EntityTeleportEvent extends EntityEvent implements ICancellableEvent {
+    private GlobalVec3 target;
+    @Deprecated(since = "1.21.1", forRemoval = true)
     protected double targetX;
+    @Deprecated(since = "1.21.1", forRemoval = true)
     protected double targetY;
+    @Deprecated(since = "1.21.1", forRemoval = true)
     protected double targetZ;
 
+    @Deprecated(since = "1.21.1", forRemoval = true)
     public EntityTeleportEvent(Entity entity, double targetX, double targetY, double targetZ) {
+        this(entity, new GlobalVec3(entity.level(), targetX, targetY, targetZ));
+    }
+
+    public EntityTeleportEvent(Entity entity, GlobalVec3 target) {
         super(entity);
-        this.targetX = targetX;
-        this.targetY = targetY;
-        this.targetZ = targetZ;
+        this.target = target;
     }
 
     public double getTargetX() {
-        return targetX;
+        return target.x;
     }
 
     public void setTargetX(double targetX) {
-        this.targetX = targetX;
+        target = target.withX(targetX);
     }
 
     public double getTargetY() {
-        return targetY;
+        return target.y;
     }
 
     public void setTargetY(double targetY) {
-        this.targetY = targetY;
+        target = target.withY(targetY);
     }
 
     public double getTargetZ() {
-        return targetZ;
+        return target.z;
     }
 
     public void setTargetZ(double targetZ) {
-        this.targetZ = targetZ;
+        target = target.withZ(targetZ);
+    }
+
+    public Level getTargetLevel() {
+        return target.level;
+    }
+
+    @ApiStatus.Internal
+    // Casting helper. Backup case shouldn't ever happen, see protection in setTargetLevel().
+    public ServerLevel getTargetServerlevelOr(ServerLevel backup) {
```

Comment:
The `level` part of the method name needs to be capitalized:
```suggestion
    public ServerLevel getTargetServerLevelOr(ServerLevel backup) {
```

Also, if the comment is to be believed that it will never reach the fallback case, why does it exist? In what possible situation could `target.level` not be a `ServerLevel`?

Reply:
No, with just the code as it is here and now. But the event classes can be subclassed and changed in the future. And what would happen then would be a class cast exception (where the event is fired) that is far removed from the code that causes it (an even subscriber)—an impossible-to-debug game crash. While it's not ideal either, I much prefer one mod's functionality silently not working. That at least gives one the chance to know which mod is doing "bad things".

And as I commented above, in the next breaking change period, changing all Level to ServerLevel would be a good idea. Then this helper wouldn't be needed anymore anyway.